File: app/src/main/java/closer/vlllage/com/closer/handler/group/GroupActionUpgradeHandler.java
package closer.vlllage.com.closer.handler.group;

import closer.vlllage.com.closer.handler.data.ApiHandler;
import closer.vlllage.com.closer.handler.helpers.ApplicationHandler;
import closer.vlllage.com.closer.handler.helpers.CameraHandler;
import closer.vlllage.com.closer.handler.helpers.DefaultAlerts;
import closer.vlllage.com.closer.handler.helpers.DisposableHandler;
import closer.vlllage.com.closer.handler.media.MediaHandler;
import closer.vlllage.com.closer.pool.PoolMember;
import closer.vlllage.com.closer.store.StoreHandler;
import closer.vlllage.com.closer.store.models.GroupAction;

public class GroupActionUpgradeHandler extends PoolMember {

    public void setPhotoFromMedia(GroupAction groupAction) {
        $(MediaHandler.class).getPhoto(photoUri -> $(PhotoUploadGroupMessageHandler.class).upload(photoUri, photoId -> {
            String photo = $(PhotoUploadGroupMessageHandler.class).getPhotoPathFromId(photoId);
            $(ApplicationHandler.class).getApp().$(DisposableHandler.class).add($(ApiHandler.class)
                    .setGroupActionPhoto(groupAction.getId(), photo).subscribe(
                            successResult -> {
                                if (successResult.success) {
                                    groupAction.setPhoto(photo);
                                    $(StoreHandler.class).getStore().box(GroupAction.class).put(groupAction);
                                } else {
                                    $(DefaultAlerts.class).thatDidntWork();
                                }
                            },
                            error -> $(DefaultAlerts.class).thatDidntWork()
                    ));
        }));
    }

    public void setPhotoFromCamera(GroupAction groupAction) {
        $(CameraHandler.class).showCamera((photoUri -> {
            $(PhotoUploadGroupMessageHandler.class).upload(photoUri, photoId -> {
                String photo = $(PhotoUploadGroupMessageHandler.class).getPhotoPathFromId(photoId);
                $(ApplicationHandler.class).getApp().$(DisposableHandler.class).add($(ApiHandler.class)
                        .setGroupActionPhoto(groupAction.getId(), photo).subscribe(
                        successResult -> {
                            if (successResult.success) {
                                groupAction.setPhoto(photo);
                                $(StoreHandler.class).getStore().box(GroupAction.class).put(groupAction);
                            } else {
                                $(DefaultAlerts.class).thatDidntWork();
                            }
                        },
                        error -> $(DefaultAlerts.class).thatDidntWork()
                ));
            });
        }));
    }
}

File: app/src/main/java/closer/vlllage/com/closer/handler/group/GroupActionUpgradeHandler.kt
package closer.vlllage.com.closer.handler.group

import closer.vlllage.com.closer.R
import closer.vlllage.com.closer.handler.data.ApiHandler
import closer.vlllage.com.closer.handler.helpers.*
import closer.vlllage.com.closer.handler.media.MediaHandler
import closer.vlllage.com.closer.store.StoreHandler
import closer.vlllage.com.closer.store.models.GroupAction
import com.google.gson.JsonArray
import com.queatz.on.On

class GroupActionUpgradeHandler constructor(private val on: On) {

    fun setPhotoFromMedia(groupAction: GroupAction) {
        on<MediaHandler>().getPhoto { photoUri ->
            on<PhotoUploadGroupMessageHandler>().upload(photoUri) { photoId ->
                val photo = on<PhotoUploadGroupMessageHandler>().getPhotoPathFromId(photoId)
                on<ApplicationHandler>().app.on<DisposableHandler>().add(on<ApiHandler>()
                        .setGroupActionPhoto(groupAction.id!!, photo).subscribe(
                                { successResult ->
                                    if (successResult.success) {
                                        groupAction.photo = photo
                                        on<StoreHandler>().store.box(GroupAction::class).put(groupAction)
                                    } else {
                                        on<DefaultAlerts>().thatDidntWork()
                                    }
                                },
                                { on<DefaultAlerts>().thatDidntWork() }
                        ))
            }
        }
    }

    fun setPhotoFromCamera(groupAction: GroupAction) {
        on<CameraHandler>().showCamera { photoUri ->
            on<PhotoUploadGroupMessageHandler>().upload(photoUri!!) { photoId ->
                val photo = on<PhotoUploadGroupMessageHandler>().getPhotoPathFromId(photoId)
                on<ApplicationHandler>().app.on<DisposableHandler>().add(on<ApiHandler>()
                        .setGroupActionPhoto(groupAction.id!!, photo).subscribe(
                                { successResult ->
                                    if (successResult.success) {
                                        groupAction.photo = photo
                                        on<StoreHandler>().store.box(GroupAction::class).put(groupAction)
                                    } else {
                                        on<DefaultAlerts>().thatDidntWork()
                                    }
                                },
                                { on<DefaultAlerts>().thatDidntWork() }
                        ))
            }
        }
    }

    fun setFlow(groupAction: GroupAction, flow: JsonArray) {
        val flowAsString = on<JsonHandler>().to(flow)
        on<DisposableHandler>().add(on<ApiHandler>().updateGroupActionFlow(groupAction.id!!, flowAsString).subscribe({
            if (it.success) {
                on<ToastHandler>().show(on<ResourcesHandler>().resources.getString(R.string.flow_updated))
                groupAction.flow = flowAsString
                on<StoreHandler>().store.box(GroupAction::class).put(groupAction)
            } else {
                on<DefaultAlerts>().thatDidntWork()
            }
        }, {
            on<DefaultAlerts>().thatDidntWork()
        }))
    }
}
